python/pyspark/sql/tests/test_connect_compatibility.py

```diff
@@ -60,16 +60,23 @@
     from pyspark.sql.connect.streaming.readwriter import DataStreamReader as ConnectDataStreamReader
     from pyspark.sql.connect.streaming.readwriter import DataStreamWriter as ConnectDataStreamWriter
 
+SPARK_CONNECT_ONLY = "spark_connect_only"
+
 
 class ConnectCompatibilityTestsMixin:
     def get_public_methods(self, cls):
         """Get public methods of a class."""
-        return {
-            name: method
-            for name, method in inspect.getmembers(cls)
-            if (inspect.isfunction(method) or isinstance(method, functools._lru_cache_wrapper))
-            and not name.startswith("_")
-        }
+        methods = {}
+        for name, method in inspect.getmembers(cls):
+            if (
+                inspect.isfunction(method) or isinstance(method, functools._lru_cache_wrapper)
+            ) and not name.startswith("_"):
+                source_lines = inspect.getsource(method).upper()
+                if "ONLY_SUPPORTED_WITH_SPARK_CONNECT" in source_lines:
```

Comment:
This is too flaky.

Reply:
Let's probably don't do this for now.

Reply:
Agree that it's flaky but I believe this is currently the only way to check functions that only supported from connect.
Should we just close it? Also cc @hvanhovell

Reply:
I don't think we should rely on reading/checking the code itself in the test. This isn't what we meant for signature comparison / compatibility test.

Reply:
How about using a new decorator `@SparkConnectOnly` which doesn't suffer from flakiness? So instead of having a list in `test_spark_session_compatibility` we now spread the items to their declaration sites.
That would make people more aware of adding such a decorator when needed.

Reply:
> How about using a new decorator @SparkConnectOnly

It sounds like a reasonable suggestion.
Let me create a separate PR and will ping you guys when it's ready.

Reply:
@xupefei just applied the comments. Could you review again when you find some time? Thanks!

Also cc @HyukjinKwon now we apply the `@spark_connect_only` decorator for APIs that are only supported with Spark Connect and use it from the testing instead of string comparison.